source/extensions/filters/http/connect_grpc_bridge/end_stream_response.cc

```diff
@@ -0,0 +1,194 @@
+#include "source/extensions/filters/http/connect_grpc_bridge/end_stream_response.h"
+
+#include "source/common/json/json_sanitizer.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ConnectGrpcBridge {
+
+namespace {
+
+using WellKnownGrpcStatus = Grpc::Status::WellKnownGrpcStatus;
+
+constexpr absl::string_view JsonCloseQuoteBrace = "\"}";
+constexpr absl::string_view JsonErrorTag = "\"error\":";
+constexpr absl::string_view JsonErrorCodeTag = "{\"code\":\"";
+constexpr absl::string_view JsonErrorMessageTag = "\",\"message\":\"";
+constexpr absl::string_view JsonErrorDetailsTypeTag = "\",\"details\":[{\"type\":\"";
+constexpr absl::string_view JsonErrorDetailNextTypeTag = "\"},{\"type\":\"";
+constexpr absl::string_view JsonErrorDetailValueTag = "\",\"value\":\"";
+constexpr absl::string_view JsonErrorDetailsClose = "\"}]}";
+constexpr absl::string_view JsonMetadataTag = "\"metadata\":{";
+
+/**
+ * Returns the appropriate error status code for a given gRPC status.
+ *
+ * https://connect.build/docs/protocol#error-codes
+ */
+absl::string_view statusCodeToString(std::string& buffer, const Grpc::Status::GrpcStatus status) {
+  switch (status) {
+  case WellKnownGrpcStatus::Canceled:
+    return "canceled";
+  case WellKnownGrpcStatus::Unknown:
+    return "unknown";
+  case WellKnownGrpcStatus::InvalidArgument:
+    return "invalid_argument";
+  case WellKnownGrpcStatus::DeadlineExceeded:
+    return "deadline_exceeded";
+  case WellKnownGrpcStatus::NotFound:
+    return "not_found";
+  case WellKnownGrpcStatus::AlreadyExists:
+    return "already_exists";
+  case WellKnownGrpcStatus::PermissionDenied:
+    return "permission_denied";
+  case WellKnownGrpcStatus::ResourceExhausted:
+    return "resource_exhausted";
+  case WellKnownGrpcStatus::FailedPrecondition:
+    return "failed_precondition";
+  case WellKnownGrpcStatus::Aborted:
+    return "aborted";
+  case WellKnownGrpcStatus::OutOfRange:
+    return "out_of_range";
+  case WellKnownGrpcStatus::Unimplemented:
+    return "unimplemented";
+  case WellKnownGrpcStatus::Internal:
+    return "internal";
+  case WellKnownGrpcStatus::Unavailable:
+    return "unavailable";
+  case WellKnownGrpcStatus::DataLoss:
+    return "data_loss";
+  case WellKnownGrpcStatus::Unauthenticated:
+    return "unauthenticated";
+  default:
+    buffer = fmt::format("code_{}", status);
+    return buffer;
+  }
+}
+
+} // namespace
+
+/**
+ * Serializes a Buf Connect Error object to JSON.
+ *
+ * https://connect.build/docs/protocol#error-end-stream
+ */
+void serializeJson(const Error& error, Buffer::Instance& buffer) {
```

Comment:
I think it's fair to say that this is not an ideal way to deal with serializing JSON. I am open to rewriting this using a different approach if there is a preferable one that would work for our use case.

Reply:
After some wrangling, Proto WKT works for both this use case as well as the JSON decoding we do in stats, so it has been replaced. Now both use Proto WKT. Resolving.